File: python/be/api/steps/apikeys_steps.py
"""Steps for be-cex-apikeys.feature. Mirror of node/be/api/steps/apikeys.steps.js.
Keys are minted with the owner Bearer token and used via X-API-Key.
"""

from __future__ import annotations

from pytest_bdd import given, parsers, then, when

from be.api.venues.minicex import ApiUnreachable, MiniCex

cex = MiniCex()


def coin(n):
    return round(float(n) * 1e8)


def acct(qa, alias):
    return qa.accounts[alias]


def act(qa, fn):
    if qa.source_error:
        return
    try:
        fn()
    except ApiUnreachable as err:
        qa.source_error = str(err)


@given(parsers.parse('"{alias}" holds a "{scope}" API key'))
def holds_key(qa, alias, scope):
    def go():
        r = cex.create_key(acct(qa, alias)["token"], scope)
        qa.current_key = {"key": r["body"]["key"], "owner": acct(qa, alias)["token"]}
    act(qa, go)


@given(parsers.parse('"{alias}" holds a "{scope}" API key limited to {rate:d} per minute'))
def holds_key_rate(qa, alias, scope, rate):
    def go():
        r = cex.create_key(acct(qa, alias)["token"], scope, rate)
        qa.current_key = {"key": r["body"]["key"], "owner": acct(qa, alias)["token"]}
    act(qa, go)


@given("the key is revoked")
def key_revoked(qa):
    act(qa, lambda: cex.revoke_key(qa.current_key["owner"], qa.current_key["key"]))


@when(parsers.parse('"{alias}" mints a "{scope}" API key'))
def mints_key(qa, alias, scope):
    act(qa, lambda: setattr(qa, "api", cex.create_key(acct(qa, alias)["token"], scope)))


@when(parsers.parse("the key places a limit sell of {size:g} {base} at {price:g} {quote}"))
def key_places(qa, size, base, price, quote):
    act(qa, lambda: setattr(qa, "api", cex.request_with_key("POST", "/orders", qa.current_key["key"], {"base": base, "quote": quote, "side": "sell", "type": "limit", "price": coin(price), "size": coin(size)})))


@when(parsers.parse("an unknown key places a limit sell of {size:g} {base} at {price:g} {quote}"))
def unknown_key_places(qa, size, base, price, quote):
    act(qa, lambda: setattr(qa, "api", cex.request_with_key("POST", "/orders", "key_nope", {"base": base, "quote": quote, "side": "sell", "type": "limit", "price": coin(price), "size": coin(size)})))


@when(parsers.parse("the key withdraws {amt:g} {asset}"))
def key_withdraws(qa, amt, asset):
    act(qa, lambda: setattr(qa, "api", cex.request_with_key("POST", "/withdraw", qa.current_key["key"], {"asset": asset, "amount": coin(amt)})))


@when(parsers.parse("the key deposits {amt:g} {asset} three times"))
def key_deposits_thrice(qa, amt, asset):
    def go():
        for _ in range(3):
            qa.api = cex.request_with_key("POST", "/deposit", qa.current_key["key"], {"asset": asset, "amount": coin(amt)})
    act(qa, go)


@when(parsers.parse('the key tries to mint a "{scope}" API key'))
def key_tries_mint(qa, scope):
    act(qa, lambda: setattr(qa, "api", cex.request_with_key("POST", "/apikeys", qa.current_key["key"], {"scope": scope})))


@when(parsers.parse('the keys of "{alias}" are read'))
def keys_read(qa, alias):
    act(qa, lambda: setattr(qa, "api", cex.keys_of(acct(qa, alias)["handle"])))


def _assert(qa, description, cond, detail):
    if qa.source_error:
        qa.unobservable(description, qa.source_error)
        return
    qa.check(description, bool(cond), detail)


@then(parsers.parse('the minted key has scope "{scope}"'))
def minted_scope(qa, scope):
    _assert(qa, "minted key scope " + scope, qa.api and qa.api["body"] and qa.api["body"].get("scope") == scope, str(qa.api["body"])[:80] if qa.api else "no response")


@then(parsers.parse("the last response status is {code:d}"))
def last_status(qa, code):
    _assert(qa, "last response status " + str(code), qa.api and qa.api["status"] == code, f"got {qa.api['status']}" if qa.api else "no response")


@then(parsers.parse('a listed key has scope "{scope}"'))
def listed_scope(qa, scope):
    found = qa.api and qa.api["body"] and any(k["scope"] == scope for k in qa.api["body"].get("keys", []))
    _assert(qa, "a listed key has scope " + scope, found, str(qa.api["body"])[:100] if qa.api else "no response")

File: node/be/api/steps/apikeys.steps.js
'use strict';

const { Given, When, Then } = require('@cucumber/cucumber');
const { MiniCex, ApiUnreachable } = require('../venues/minicex');

/**
 * Steps for features/be-cex-apikeys.feature. Keys are minted with the owner's
 * Bearer token, then used via the X-API-Key header; the scope gates what they
 * can do and the rate limit throttles them. The account/deposit steps are
 * shared from cex.steps.js; the response assertions from minicex.steps.js.
 */

const cex = new MiniCex();
const coin = (n) => Math.round(Number(n) * 1e8);
const acctOf = (world, alias) => world.accounts[alias];
const SELL = { base: 'BTC', quote: 'USD', side: 'sell', type: 'limit', price: coin(60000), size: coin(0.5) };

async function act(world, fn) {
  if (world.sourceError) return;
  try { await fn(); } catch (err) { if (err instanceof ApiUnreachable) { world.sourceError = err.message; return; } throw err; }
}

Given('{string} holds a {string} API key', async function (alias, scope) {
  await act(this, async () => { const r = await cex.createKey(acctOf(this, alias).token, scope); this.currentKey = { key: r.body.key, owner: acctOf(this, alias).token }; });
});
Given('{string} holds a {string} API key limited to {int} per minute', async function (alias, scope, rate) {
  await act(this, async () => { const r = await cex.createKey(acctOf(this, alias).token, scope, rate); this.currentKey = { key: r.body.key, owner: acctOf(this, alias).token }; });
});
Given('the key is revoked', async function () {
  await act(this, async () => { await cex.revokeKey(this.currentKey.owner, this.currentKey.key); });
});

When('{string} mints a {string} API key', async function (alias, scope) {
  await act(this, async () => { this.api = await cex.createKey(acctOf(this, alias).token, scope); });
});
When('the key places a limit sell of {float} {word} at {float} {word}', async function (size, base, price, quote) {
  await act(this, async () => { this.api = await cex.requestWithKey('POST', '/orders', this.currentKey.key, { base, quote, side: 'sell', type: 'limit', price: coin(price), size: coin(size) }); });
});
When('an unknown key places a limit sell of {float} {word} at {float} {word}', async function (size, base, price, quote) {
  await act(this, async () => { this.api = await cex.requestWithKey('POST', '/orders', 'key_nope', { base, quote, side: 'sell', type: 'limit', price: coin(price), size: coin(size) }); });
});
When('the key withdraws {float} {word}', async function (amt, asset) {
  await act(this, async () => { this.api = await cex.requestWithKey('POST', '/withdraw', this.currentKey.key, { asset, amount: coin(amt) }); });
});
When('the key deposits {float} {word} three times', async function (amt, asset) {
  await act(this, async () => { for (let i = 0; i < 3; i++) this.api = await cex.requestWithKey('POST', '/deposit', this.currentKey.key, { asset, amount: coin(amt) }); });
});
When('the key tries to mint a {string} API key', async function (scope) {
  await act(this, async () => { this.api = await cex.requestWithKey('POST', '/apikeys', this.currentKey.key, { scope }); });
});
When('the keys of {string} are read', async function (alias) {
  await act(this, async () => { this.api = await cex.keysOf(acctOf(this, alias).handle); });
});

Then('the minted key has scope {string}', function (scope) {
  const ok = this.api && this.api.body && this.api.body.scope === scope;
  if (this.sourceError) { this.unobservable('minted key scope', this.sourceError); return; }
  this.check('minted key scope ' + scope, ok, this.api ? JSON.stringify(this.api.body).slice(0, 80) : 'no response');
});
Then('the last response status is {int}', function (code) {
  if (this.sourceError) { this.unobservable('last response status', this.sourceError); return; }
  this.check('last response status ' + code, this.api && this.api.status === code, this.api ? 'got ' + this.api.status : 'no response');
});
Then('a listed key has scope {string}', function (scope) {
  if (this.sourceError) { this.unobservable('listed key scope', this.sourceError); return; }
  const found = this.api && this.api.body && (this.api.body.keys || []).some((k) => k.scope === scope);
  this.check('a listed key has scope ' + scope, found, this.api ? JSON.stringify(this.api.body).slice(0, 100) : 'no response');
});
